spec/client/ics-007-tendermint-client/README.md

```diff
@@ -184,69 +196,133 @@ function latestClientHeight(clientState: ClientState): Height {
+        verifyHeader(clientState, clientMsg)
+      case Misbehaviour:
+        verifyHeader(clientState, clientMsg.h1)
+        verifyHeader(clientState, clientMsg.h2)
+    }
+}
+```
+
+Verify validity of regular update to the Tendermint client
+
+```typescript
+verifyHeader(clientState, header) {
     // assert trusting period has not yet passed
     assert(currentTimestamp() - clientState.latestTimestamp < clientState.trustingPeriod)
     // assert header timestamp is less than trust period in the future. This should be resolved with an intermediate header.
     assert(header.timestamp - clientState.latestTimeStamp < trustingPeriod)
-    // assert header timestamp is past current timestamp
-    assert(header.timestamp > clientState.latestTimestamp)
-    // assert header height is newer than any we know
-    assert(header.height > clientState.latestHeight)
+    // trusted height revision must be the same as header revision
+    // if revisions are different, use upgrade client instead
+    // trusted height must be less than header height
+    assert(header.height.revisionNumber == header.trustedHeight.revisionNumber)
+    assert(header.height.revisionHeight > header.trustedHeight.revisionHeight)
+    // fetch the consensus state at the trusted height
+    consensusState = get("clients/{identifier}/consensusStates/{header.trustedHeight}")
+    // assert that header's trusted validator set hashes to consensus state's validator hash
+    assert(hash(header.trustedValidatorSet) == consensusState.nextValidatorsHash)
+
     // call the `verify` function
-    assert(verify(clientState.validatorSet, clientState.latestHeight, clientState.trustingPeriod, maxClockDrift, header))
-    // update validator set
-    clientState.validatorSet = header.validatorSet
-    // update latest height
-    clientState.latestHeight = header.height
-    // update latest timestamp
-    clientState.latestTimestamp = header.timestamp
+    assert(verify(
+      header.trustedValidatorSet,
+      clientState.latestHeight,
+      clientState.trustingPeriod,
+      clientState.maxClockDrift,
+      header.TendermintSignedHeader,
+    ))
+}
+```
+
+### CheckForMisbehaviour
+
+CheckForMisbehaviour will check if an update contains evidence of Misbehaviour. If the ClientMessage is a header we check for implicit evidence of misbehaviour by checking if there already exists a conflicting consensus state in the store or if the header breaks time monotonicity.
+
+```typescript
+function checkForMisbehaviour(
+  clientState: clientState,
+  clientMsg: clientMessage) => bool {
+    switch typeof(clientMsg) {
+      case Header:
+        // fetch consensusstate at header height if it exists
+        consensusState = get("clients/{identifier}/consensusStates/{header.Height}")
+        // if consensus state exists and conflicts with the header
+        // then the header is evidence of misbehaviour
+        if consensusState != nil && consensusState.commitmentRoot != header.AppHash {
```

Comment:
`header.AppHash` is not defined. Should this be `header. commitmentRoot` instead?